newrelic-agent/src/main/java/com/newrelic/agent/stats/StatsImpl.java

```diff
@@ -58,84 +62,100 @@ public void recordDataPoint(float value) {
         if (Float.isNaN(value) || Float.isInfinite(value)) {
             throw new IllegalArgumentException("Data points must be numbers");
         }
-        double sos = sumOfSquares + (value * value);
-        if (sos < sumOfSquares) {
-            throw new IllegalArgumentException("Data value " + value + " caused sum of squares to roll over");
-        }
-        if (count > 0) {
-            minValue = Math.min(value, minValue);
-        } else {
-            minValue = value;
-        }
-        count++;
-        total += value;
-        maxValue = Math.max(value, maxValue);
-        sumOfSquares = sos;
+        synchronized (lock) {
+            double sos = sumOfSquares + (value * value);
+            if (sos < sumOfSquares) {
+                throw new IllegalArgumentException("Data value " + value + " caused sum of squares to roll over");
+            }
+            if (count > 0) {
+                minValue = Math.min(value, minValue);
+            } else {
+                minValue = value;
+            }
+            count++;
+            total += value;
+            maxValue = Math.max(value, maxValue);
+            sumOfSquares = sos;
 
-        if (NewRelic.getAgent().getConfig().getValue(AgentConfigImpl.METRIC_DEBUG, AgentConfigImpl.DEFAULT_METRIC_DEBUG)) {
-            if (count < 0 || total < 0) {
-                NewRelic.incrementCounter("Supportability/StatsImpl/NegativeValue");
-                Agent.LOG.log(Level.INFO, "Invalid count {0} or total {1}", count, total);
+            if (NewRelic.getAgent().getConfig().getValue(AgentConfigImpl.METRIC_DEBUG, AgentConfigImpl.DEFAULT_METRIC_DEBUG)) {
+                if (count < 0 || total < 0) {
+                    NewRelic.incrementCounter("Supportability/StatsImpl/NegativeValue");
+                    Agent.LOG.log(Level.INFO, "Invalid count {0} or total {1}", count, total);
 
+                }
             }
         }
-
     }
 
     @Override
     public boolean hasData() {
-        return count > 0 || total > 0;
+        synchronized (lock) {
+            return count > 0 || total > 0;
+        }
     }
 
     @Override
     public void reset() {
-        count = 0;
-        total = minValue = maxValue = 0;
-        sumOfSquares = 0;
+        synchronized (lock) {
+            count = 0;
+            total = minValue = maxValue = 0;
+            sumOfSquares = 0;
+        }
     }
 
     @Override
     public float getTotal() {
-        return total;
+        synchronized (lock) {
+            return total;
```

Comment:
reading a `float` is  atomic.  if you mark `volatile`, you could avoid the locking here.

Reply:
ok, cool...played with volatile but wasn't sure.